### PR TITLE
Add System.Reflection.Emit.DynamicILInfo

### PR DESCRIPTION
**Pull request by maryamariyan**
Related to https://github.com/dotnet/corefx/issues/12055

cc: @Anipik @danmosemsft 

related corefx PR: https://github.com/dotnet/corefx/pull/34532

### CODE REVIEW

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 1]
```diff
@@ -901,8 +915,109 @@ internal byte[] LocalSignature
...
+            m_maxStackSize = maxStackSize;
+        }
+
+        [CLSCompliant(false)]
+        public unsafe void SetCode(byte* code, int codeSize, int maxStackSize)
+        {
+            if (codeSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(codeSize), SR.ArgumentOutOfRange_GenericPositive);
+            if (codeSize > 0 && code == null)
+                throw new ArgumentNullException("code");
```
Comment (Anipik) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: nit: use nameof(code)

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 2]
```diff
@@ -875,6 +875,20 @@ internal class DynamicILInfo
         private int m_methodSignature;
         #endregion
 
+        #region Constructor
+        internal DynamicILInfo(object scope, DynamicMethod method, byte[] methodSignature)
+        {
+            if (scope as DynamicScope is null)
+                throw new InvalidCastException(nameof(scope));
+            m_method = method;
```
Comment (Anipik) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: while you are here, you might wanna change the names of this variables to new convention i.e private fields starting with _

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 3]
```diff
@@ -901,8 +915,109 @@ internal byte[] LocalSignature
...
+        }
+
+        [CLSCompliant(false)]
+        public unsafe void SetExceptions(byte* exceptions, int exceptionsSize)
+        {
+            if (exceptionsSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(exceptionsSize), SR.ArgumentOutOfRange_GenericPositive);
+
+            if (exceptionsSize > 0 && exceptions == null)
+                throw new ArgumentNullException("exceptions");
```
Comment (Anipik) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: nit nameof(exceptions)

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 4]
```diff
@@ -901,8 +915,109 @@ internal byte[] LocalSignature
...
+        }
+
+        [CLSCompliant(false)]
+        public unsafe void SetLocalSignature(byte* localSignature, int signatureSize)
+        {
+            if (signatureSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(signatureSize), SR.ArgumentOutOfRange_GenericPositive);
+
+            if (signatureSize > 0 && localSignature == null)
+                throw new ArgumentNullException("localSignature");
```
Comment (Anipik) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: nit use nameof(localSignature)

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs (hunk 2 above)
Comment (maryamariyan) [reply src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: I tried doing that, but there's a lot of these private fields with m_ in both DynamicILGenerator and DynamicILInfo. And since they already exist in code, I think it's better to leave as is.

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 5]
```diff
@@ -901,8 +913,109 @@ internal byte[] LocalSignature
...
+        [CLSCompliant(false)]
+        public unsafe void SetCode(byte* code, int codeSize, int maxStackSize)
+        {
+            if (codeSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(codeSize), SR.ArgumentOutOfRange_GenericPositive);
+            if (codeSize > 0 && code == null)
+                throw new ArgumentNullException(nameof(code));
+
+            m_code = new byte[codeSize];
+            for (int i = 0; i < codeSize; i++)
```
Comment (benaadams) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: nit use `.Length` to elide the bounds check
```
for (int i = 0; i < m_code.Length; i++)
```
or better  use local var to be sure
```csharp
byte[] codes = new byte[codeSize];
for (int i = 0; i < codes.Length; i++)
{
    codes[i] = *code;
    code++;
}

m_code = codes;
```

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 6]
```diff
@@ -901,8 +913,109 @@ internal byte[] LocalSignature
...
+        {
+            if (exceptionsSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(exceptionsSize), SR.ArgumentOutOfRange_GenericPositive);
+
+            if (exceptionsSize > 0 && exceptions == null)
+                throw new ArgumentNullException(nameof(exceptions));
+
+            m_exceptions = new byte[exceptionsSize];
+
+            for (int i = 0; i < exceptionsSize; i++)
```
Comment (benaadams) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: Same here

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 7]
```diff
@@ -901,8 +913,109 @@ internal byte[] LocalSignature
...
+        public unsafe void SetLocalSignature(byte* localSignature, int signatureSize)
+        {
+            if (signatureSize < 0)
+                throw new ArgumentOutOfRangeException(nameof(signatureSize), SR.ArgumentOutOfRange_GenericPositive);
+
+            if (signatureSize > 0 && localSignature == null)
+                throw new ArgumentNullException(nameof(localSignature));
+
+            m_localSignature = new byte[signatureSize];
+            for (int i = 0; i < signatureSize; i++)
```
Comment (benaadams) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: Same here

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: Couldn't we just use span?
```C#
m_code = new Span<byte>(code, codeSize).ToArray();
```

File: src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs [hunk 8]
```diff
@@ -875,6 +875,18 @@ internal class DynamicILInfo
         private int m_methodSignature;
         #endregion
 
+        #region Constructor
+        internal DynamicILInfo(object scope, DynamicMethod method, byte[] methodSignature)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: Why did you change the first argument type from DynamicScope to object? It adds unnecessary cast below.

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: BTW: The DynamicScope argument looks useless. DynamicScope is internal detail of DynamicILInfo. You can just do `new DynamicScope` inside the constructor. It does not need to be passed in from outside.

Comment (maryamariyan) [reply src/System.Private.CoreLib/src/System/Reflection/Emit/DynamicILGenerator.cs]: Yes, the original reason was because I didn't want to have to expose that type.  Best is to instantiate it in ctor as you suggested.
